### ISSUE TITLE
[Network] App-Gateway: Allow creation of health probe at create time

### ISSUE THREAD

**Issue by andymoe** [Azure/azure-cli#1500]
There is not way to create a health check prob when creating an application gateway. This makes it impossible to automate the creation of a functional app gateway that also terminates TLS traffic.

Please see for the same issue in the other cli scheduled to be released dec 14th: https://github.com/Azure/azure-xplat-cli/issues/3283

**Comment by tjprescott** [comment]
The `az network application-gateway http-settings create` command lets you specify a health probe. Does this not work for your scenario?

**Comment by andymoe** [comment]
Hey @tjprescott I believe the state that an application gateway is left in when it's initially created makes for TLS termination makes it impossible to update the existing http-settings (httpSettings01) with that command. 

I think you really need to be able to specify a probe when the app gateway is created.  The reason for this is that default http-settings is attached to a rule and the rule is attached to listener and the listener to a public ip. This is all configured by default to so to attach a new http-setting with a probe these all have to be destroyed and the dependencies don't allow for that. Am I missing something obvious?

**Comment by tjprescott** [comment]
Hi @andymoe can you enable your scenario via the portal? If so, what steps do you take?

Can you try the following to update the HttpSettings with your custom probe:
`az network application-gateway update -g {resource-group-name} -n {app-gateway-name} --set backendHttpSettingsCollection[0].probe={health-probe-id}`

The reason we don't create the probe is because the portal doesn't and we tend to expose the portal's functionality. However, the portal lets you add it after the fact. The CLI doesn't have an `application-gateway http-settings update` command, but you should be able to do this through the root application-gateway generic update.

Either way, I'll look at adding this to the create step since application gateways take so long to create, but it would be good to know if this works in your scenario.

**Comment by andymoe** [comment]
Yes, I can enable the scenario via the portal. All that is required is to create an app gateway that terminates SSL with a backend pool and traffic to the pool via http. I then create a custom health check probe pointing to /healthcheck or even / over http . The probe must be provided otherwise the app gateway seems to try to healthcheck to / over https but the backend pool of servers expects to traffic over http (since I want to offload https traffic and config) and so the app gateway gets unhealthy and 502s (or maybe 503 can't quite remember).

Here is what the commands might look like to set this all up:

> az network public-ip create \
>  --name "public_gateway_ip" \
>  --resource-group "my_group" \
>  --location "westus"``


Note: In my script I actually use the old xsplat CLI for this portion because of problems with the `--servers` arg. That's for another issue report though.
>  BACKEND_POOL_IP=172.17.0.4
> az network application-gateway create \
>  --name "my_app_gateway" \
>  --resource-group "my_group" \
>  --location "westus" \
>  --vnet-name "my_vnet" \
>  --subnet-name "my_gateway_subnet" \
>  --servers "${BACKEND_POOL_IP}" \
>  --capacity 2 \
>  --sku-name Standard_Large \
>  --cert-file "my_cert.pfx" \
>  --cert-password "a_cert_password" \
>  --http-listener-protocol https \
>  --frontend-port 443 \
>  --public-ip-name "$(PUBLIC_GATEWAY_IP}"

// this custom probe is then manually attached in the Azure website UI.
>   az network application-gateway probe create \
>  --gateway-name "my_app_gateway" \
>  --resource-group "my_group" \
>  --name "${GATEWAY_PROBE_NAME}" \
>  --host "${PRIVATE_API_IP}" \
>  --path "${API_HEALTH_CHECK_PATH}" \
>  --protocol Http \
>  --interval 30 \
>  --threshold 8 \
>  --timeout 120

Still have not had time to try the command you recommended. I'll try to incorporate it into my deploy script tomorrow... my feedback there is that the commands that take `--set` are confusing and not well documented. 

Let me know if you have any other questions. Thanks again!

**Comment by tjprescott** [comment]
> Note: In my script I actually use the old xplat CLI for this portion because of problems with the --servers flag. That's for another issue report though.

Can you report that issue please? I plan on addressing all the application-gateway fixes in a PR before the holiday, so if there's another issue, I would like to address it in the same PR.

One thing I would note that is that if you are doing:
`--servers "${BACKEND_POOL_IP}"` 

And trying to use a space separated list of servers, that would look to the CLI like `--servers "10.0.0.0 10.0.0.1"` which *will* fail because it is interpreted as a single string with spaces.

**Comment by andymoe** [comment]
@tjprescott, ok, I'll try to reproduce the `--servers` issue. would you expect me to be able to pass a single quoted address like `--servers "10.0.0.1"` what about `--servers "10.0.0.1, 10.0.0.1"` the other old `azure` client says to use a comma separated like but this one just says `The list of IP addresses or DNS names corresponding to backend servers.`

**Comment by tjprescott** [comment]
@andymoe
**These Will Work**
`--servers "10.0.0.1"`
`--servers "10.0.0.1" "10.0.0.1"`
`--servers 10.0.0.1 10.0.0.1`
`--servers 10.0.0.1 myserver.name.foo`

**These Will NOT Work**
`--servers "10.0.0.1, 10.0.0.1"`
`--servers "10.0.0.1 10.0.0.1"`
`--servers 10.0.0.1,10.0.0.2`

When I work on application gateway I'll clarify the help to say a space-separated list of things.

Quotes are pretty much only need in the CLI if you want a string value that has spaces, such as:
`--tags mytag="some value with spaces"`

**Comment by andymoe** [comment]
@tjprescott, thanks. Very good chance we just messed that up since we were trying to switch from the other CLI. I'll confirm.

**Comment by andymoe** [comment]
So here is what happens when I try to update the backendHttpSettingsCollection directly... Any thoughts would be appreciated:

```
az network application-gateway update --resource-group my_group --name my_app_gateway_name \
--set backendHttpSettingsCollection[0].probe=/subscriptions/*****/resourceGroups/my_group/providers/Microsoft.Network/applicationGateways/my_app_gateway_name/probes/my_probe_name


string indices must be integers, not str
Traceback (most recent call last):
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/azure/cli/main.py", line 34, in main
    cmd_result = APPLICATION.execute(args)
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/azure/cli/core/application.py", line 139, in execute
    result = expanded_arg.func(params)
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/azure/cli/core/commands/arm.py", line 293, in handler
    opres = setter(client, **getterargs) if client else setter(**getterargs)
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/azure/mgmt/network/operations/application_gateways_operations.py", line 223, in create_or_update
    body_content = self._serialize.body(parameters, 'ApplicationGateway')
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/msrest/serialization.py", line 303, in body
    data = _convert_to_datatype(data, data_type, self.dependencies)
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/msrest/serialization.py", line 171, in _convert_to_datatype
    getattr(data, attr), map['type'], localtypes))
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/msrest/serialization.py", line 150, in _convert_to_datatype
    param, data_type[1:-1], localtypes) for param in data
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/msrest/serialization.py", line 171, in _convert_to_datatype
    getattr(data, attr), map['type'], localtypes))
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/msrest/serialization.py", line 164, in _convert_to_datatype
    localtypes) for key in data
  File "/home/jumpboxadmin/lib/azure-cli/local/lib/python2.7/site-packages/msrest/serialization.py", line 164, in <dictcomp>
    localtypes) for key in data
TypeError: string indices must be integers, not str

```

**Comment by tjprescott** [comment]
@andymoe here's what worked for me (beware--it's ugly!)

`network application-gateway update -g rg -n ag --set backendHttpSettingsCollection[0].probe={\"id\":\"/subscriptions/****/resourceGroups/rg/providers/Microsoft.Network/applicationGateways/ag/probes/probe1\"}`

Within the CLI, the object updates properly, but then when msrest tries to serialize it, it chokes. I think it is because the `probe` property of the `backend_http_settings_collection` is listed as a "SubResource" rather than a "Probe", but that will let you set the probe. Whether that unblocks your scenario or not I'd be curious to know, but I already plan on adding proper arguments to update these kinds of properties, and will likely allow creation of the probe at the same time as the gateway.

**Comment by andymoe** [comment]
Cool, thanks @tjprescott  well it *is* ugly but it does seem to attach the custom probe. The downside is it makes the Web UI just return very basic info instead of the full UI for the app gateway now.

**Comment by tjprescott** [comment]
@andymoe 

>  The downside is it makes the Web UI just return very basic info instead of the full UI for the app gateway now

I'm not sure what you mean here--the UI in the portal?

**Comment by andymoe** [comment]
@tjprescott After I ran the script to automatically add the probes the UI in the portal did not show anything under the "settings" heading for any of the app gateways the script had touched. Funny thing is this morning I logged in again and everything looks normal again. I'll run the scripts and take a screen shot if I'm able to reproduce.

**Comment by tjprescott** [comment]
Related fix in Xplat https://github.com/Azure/azure-xplat-cli/pull/3315

**Comment by tjprescott** [comment]
> network application-gateway update -g rg -n ag --set backendHttpSettingsCollection[0].probe={\"id\":\"/subscriptions/****/resourceGroups/rg/providers/Microsoft.Network/applicationGateways/ag/probes/probe1\"}
 
@andymoe actually, this syntax works:

`network application-gateway update -g rg -n ag --set backendHttpSettingsCollection[0].probe.id=/subscriptions/****/resourceGroups/rg/providers/Microsoft.Network/applicationGateways/ag/probes/probe1`

**Comment by andymoe** [comment]
Hey @tjprescott I finally was able to reproduce the issue re --servers using the az CLI to create an app gateway. Sorry it took so long. Separate issue filed here: https://github.com/Azure/azure-cli/issues/1619

**Comment by tjprescott** [comment]
The bug aspects of this have been addressed in other PRs so that scenario is unblocked. There is ongoing work that may change the creation process for app-gateway so I will put this enhancement on the backlog for now until that work is complete.

**Comment by andymoe** [comment]
Sounds good @tjprescott - thanks again for all your help and work over the past several weeks.

**Comment by tjprescott** [comment]
A strongly-typed update command has been added for `application-gateway http-settings`, so I believe this should be sufficient. The use of generic update for this scenario is no longer necessary. Closing.
